runtime/src/slots.rs

```diff
@@ -0,0 +1,1425 @@
+				}
+			}
+		}
+		None
+	}
+
+	/// Returns the current lease period.
+	fn lease_period_index() -> LeasePeriodOf<T> {
+		(<system::Module<T>>::block_number() / T::LeasePeriod::get()).into()
+	}
+
+	/// Some when the auction's end is known (with the end block number). None if it is unknown.
+	/// If `Some` then the block number must be at most the previous block and at least the
+	/// previous block minus `T::EndingPeriod::get()`.
+	///
+	/// This mutates the state, cleaning up `AuctionInfo` and `Winning` in the case of an auction
+	/// ending. An immediately subsequent call with the same argument will always return `None`.
+	fn check_auction_end(now: T::BlockNumber) -> Option<(WinningData<T>, LeasePeriodOf<T>)> {
+		if let Some((lease_period_index, early_end)) = <AuctionInfo<T>>::get() {
+			if early_end + T::EndingPeriod::get() == now {
+				// Just ended!
+				let ending_period = T::EndingPeriod::get();
+				let offset = T::BlockNumber::decode(&mut<system::Module<T>>::random_seed().as_ref())
+					.expect("secure hashes always bigger than block numbers; qed") % ending_period;
+				let res = <Winning<T>>::get(offset).unwrap_or_default();
+				let mut i = T::BlockNumber::zero();
+				while i < ending_period {
+					<Winning<T>>::remove(i);
+					i += One::one();
+				}
+				<AuctionInfo<T>>::kill();
+				return Some((res, lease_period_index))
+			}
+		}
+		None
+	}
+
+	/// Auction just ended. We have the current lease period, the auction's lease period (which
+	/// is guaranteed to be at least the current period) and the bidders that were winning each
+	/// range at the time of the auction's close.
+	fn manage_auction_end(
+		lease_period_index: LeasePeriodOf<T>,
+		auction_lease_period_index: LeasePeriodOf<T>,
+		winning_ranges: WinningData<T>,
+	) {
+		// First, unreserve all amounts that were reserved for the bids. We will later deduct the
+		// amounts from the bidders that ended up being assigned the slot so there's no need to
+		// special-case them here.
+		for (bidder, _) in winning_ranges.iter().filter_map(|x| x.as_ref()) {
+			if let Some(amount) = <ReservedAmounts<T>>::take(bidder) {
+				T::Currency::unreserve(&bidder.funding_account(), amount);
+			}
+		}
+
+		// Next, calculate the winning combination of slots and thus the final winners of the
+		// auction.
+		let winners = Self::calculate_winners(winning_ranges, T::Parachains::new_id);
+
+		Self::deposit_event(RawEvent::AuctionClosed(Self::auction_counter()));
+
+		// Go through those winners and deduct their bid, updating our table of deposits
+		// accordingly.
+		for (maybe_new_deploy, para_id, amount, range) in winners.into_iter() {
+			match maybe_new_deploy {
+				Some(bidder) => {
+					// For new deployments we ensure the full amount is deducted. This should always
+					// succeed as we just unreserved the same amount above.
+					if T::Currency::withdraw(
+						&bidder.who,
+						amount,
+						WithdrawReason::Fee,
+						ExistenceRequirement::AllowDeath
```

Comment:
Should there be some sort of warning to users that if they over-bid, their account can be cleaned up? Maybe not in the code here, but in documentation.

Reply:
no:

1. it's not meant to be used by bidders directly - this is just backend code that a UI should be using.
2. the convention already exists. it's no different to other functions like `transfer` and the user should be clear that anything they do which could result in the balance dropping below the existential deposit can result in the account being lost. wallets should account for this.